Implementation/48289 add automatic project folder management settings for storages

akabiru opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/48289

Let OpenProject create folders per project automatically. This is recommended as it ensures that every team member always has the correct access permissions.

- [x] Define UI blueprint
  - [x] Update from checkbox to `spot-switch`
- [x] Redirect from oauth client form (❓ _Feature flag_)
  - [x] _Optional_ scope configuration in nested controller
- [x] _FIXME:_ _update html title_
- [x] Persistence rules
  - [x] Default values: toggle is on, require app password
  - [x] username should always be "OpenProject" if storage provider type is nextcloud
  - [x] Clear `application`_{username,password}` on `automatically_managed` `false`
  - [x] Only show managed project folders form when there were no pre-existing values (during edit of oauth credentials)
- [x] Display errors on form validation failure
- [x] Add managed project folders index & edit views
   -  🚧 Add `applicationPassword` validation ([Separate WP](https://community.openproject.org/projects/openproject/boards/3944/details/48569/overview))
   - [x] On edit, show password filled status as masked placeholder
- [x] Add API
   - [x] Add new storage representer parsing spec
   - [x] Ensure `applicationPassword` is not rendered in storage representer rendering
   - [x] Add request specs
- [x] Move models logic for nextcloud to `Storages::NextcloudStorge`
<details> <summary> Screenshots 🔽 </summary>

![Screenshot 2023-06-14 at 11 58 42 AM](https://github.com/opf/openproject/assets/17295175/21f98aea-b6eb-4430-b556-02a5244a3aa2)
![Screenshot 2023-06-14 at 11 56 31 AM](https://github.com/opf/openproject/assets/17295175/fabdf402-91dd-47b9-a3d6-b1c0acfcb21f)
![Screenshot 2023-06-14 at 11 55 55 AM](https://github.com/opf/openproject/assets/17295175/d64f3976-02a8-4be8-947d-b93f546037d5)
![Screenshot 2023-06-14 at 11 55 44 AM](https://github.com/opf/openproject/assets/17295175/724c9e44-b1ce-4521-9640-14750ac4f7d0)
![Screenshot 2023-06-14 at 11 55 16 AM](https://github.com/opf/openproject/assets/17295175/e586c41f-8768-4847-bc58-9289c0c9f628)


</details>

akabiru committed:
[#48289] Add automatic project folder management settings for storages

https://community.openproject.org/work_packages/48289

akabiru committed:
feat[#48289] Define UI blueprint

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Organise UI templates better

https://community.openproject.org/work_packages/48289

akabiru committed:
tests[#48289] Add Storage provider fields specs

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Update rename to `automatically_managed(?)`

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] decouple contract and attribute setters from model

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Add missing validation error messages notice

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Minor touch-ups

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Replace `File.join` with `URI.join` for semantics

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Add HTML Input[value] to spot switch for semantic form submission

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Cover case when checkbox is unchecked and no value is submitted

https://community.openproject.org/work_packages/48289

akabiru committed:
feat[#48289] Define `ManagedProjectFolderFormController` for toggle behaviour

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Update HTML title

https://community.openproject.org/work_packages/48289

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

akabiru committed:
chore[#48289] Reuse storages update service

The side effect of `::OAuth::PersistApplicationService` is safe as it
ensures the `name` and `redirect_uri` are always up to date.

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Redefine implementation at provider layer

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Rename labels to `Automatically managed project folders`

https://community.openproject.org/work_packages/48289

akabiru committed:
tests[#48289] Define missing `StoragesRepresenter` parsing spec

https://community.openproject.org/work_packages/48289

akabiru committed:
feat[#48289] Define application password representation

https://community.openproject.org/work_packages/48289

akabiru committed:
tests[#48289] Ensure `applicationPassword` is never rendered via representer

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Set default username correctly

`represented` is not a `Storages::Storage` rather an `API::ParserStruct`

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Set attributes centrally in service

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Check for `application_password` presence

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Resolve contract form validation regression

https://community.openproject.org/work_packages/48289

akabiru committed:
fix[#48289] Do not overwrite application credentials during first time creation

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Define `Storage#automatic_management_unspecified?` to clarify nil checks

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Define index and edit flows with feature specs

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Add note on `SetProviderFieldsAttributesService` usage

https://community.openproject.org/work_packages/48289

wielinde reviewed (changes requested):
@akabiru I haven't yet fully worked my way through the code. However, I already give you feedback on what I have observed:
- On my dev machine, I had storages configured before applying your branch. They do not yet know anything about app passwords and modes etc. When I edit on of those, the button for adding the missing configuration is "+ Create". We didn't specify that case in Figma. I think we should call int "Configure ..." or something link that (@marcalcobe what do you think?)
- Then, after pressing the button, I would have expected that the switch was set to true (as that is the default for not present configurations). However, for me it shows itself as off.

akabiru committed:
chore[#48289]: Update UI for automatically managed configuration

1. When there was no storage before, let the label read "configure"
2. Redirect to new for defaults to kick in

See: https://github.com/opf/openproject/pull/12721#pullrequestreview-1479195251

akabiru committed:
chore[#48289] UI Update: shorten instruction

https://community.openproject.org/work_packages/48289

wielinde commented on modules/storages/app/views/storages/admin/storages/show_managed_project_folders.html.erb:
```diff
@@ -0,0 +1,102 @@
+<%= error_messages_for_contract @storage, @errors %>
+
+<%= labelled_tabular_form_for @storage, url: admin_settings_storage_managed_project_folders_path(@storage),
+                                        method: :patch, data: {
```
line break behind the comma, right?

akabiru committed:
chore[#48289] Add new line for readability

https://community.openproject.org/work_packages/48289

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

wielinde commented on modules/storages/app/models/storages/storage.rb:
Probably most of the changes are Nextcloud specific and belong to `NextcloudStorage`

wielinde commented on modules/storages/spec/lib/api/v3/storages/storages_representer_parsing_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        expect(parsed).to have_attributes(name: 'Nextcloud Local', host: storage.host,
+                                          provider_type: "Storages::NextcloudStorage")
+
+        aggregate_failures 'does not parse applicationPassword' do
```
```suggestion
        aggregate_failures 'does not parse hasApplicationPassword' do
```

wielinde commented on modules/storages/spec/lib/api/v3/storages/storages_representer_parsing_spec.rb:
```diff
@@ -0,0 +1,103 @@
+
+      context 'with applicationPassword blank' do
+        let(:parsed_hash) do
+          super().merge(
```
```suggestion
          super.merge(
```

akabiru replied:
Just found out implicit argument passing is not supported in meta context 🎓 

```ruby
super.merge(
         "applicationPassword" => ''
       )

     RuntimeError:
       implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.
```

wielinde commented on modules/storages/app/controllers/storages/admin/managed_project_folders_controller.rb:
```diff
@@ -0,0 +1,138 @@
+  before_action :require_admin
+
+  # specify which model #find_model_object should look up
+  model_object Storages::Storage
```
```suggestion
  model_object Storages::NextcloudStorage
```

wielinde commented on modules/storages/app/controllers/storages/admin/managed_project_folders_controller.rb:
```diff
@@ -0,0 +1,138 @@
+                     contract_class: ::Storages::Storages::BaseContract)
+                .call
+                .result
+    render '/storages/admin/storages/show_managed_project_folders'
```
```suggestion
    render '/storages/admin/storages/edit_managed_project_folders'
```

wielinde commented on modules/storages/app/controllers/storages/admin/managed_project_folders_controller.rb:
```diff
@@ -0,0 +1,138 @@
+  # Used by: The StoragesController#edit, when user wants to update application credentials.
+  # Called by: Global app/config/routes.rb to serve Web page
+  def edit
+    render '/storages/admin/storages/show_managed_project_folders'
```
```suggestion
    render '/storages/admin/storages/edit_managed_project_folders'
```

wielinde commented on modules/storages/app/controllers/storages/admin/managed_project_folders_controller.rb:
```diff
@@ -0,0 +1,138 @@
+      redirect_to admin_settings_storage_path(@storage)
+    else
+      @errors = service_result.errors
+      render '/storages/admin/storages/show_managed_project_folders'
```
```suggestion
      render '/storages/admin/storages/edit_managed_project_folders'
```

wielinde reviewed (changes requested):
Looks all very good. Find minor naming comments inline.

One slightly bigger topic is moving Nextcloud specifics from the `Storage` model to `NextcloudStorage` model.

akabiru committed:
tests[#12721]: specify test subject is `hasApplicationPassword`

Co-authored-by: Wieland Lindenthal <w.lindenthal@openproject.com>

akabiru committed:
chore[#48289] Rename controller to `AutomaticallyManagedProjectFolders`

https://community.openproject.org/work_packages/48289

akabiru committed:
chore[#48289] Adopt `NextcloudStorage` model

https://community.openproject.org/work_packages/48289

akabiru committed:
feat[#48289] Add `palkan/stored_attribute` for enhance typecasting

[palkan/stored_attribute](https://github.com/palkan/store_attribute)

> ActiveRecord extension which adds typecasting to store accessors

https://evilmartians.com/opensource/store-attribute
https://community.openproject.org/work_packages/48289

akabiru commented on modules/storages/app/models/storages/storage.rb:
```diff
@@ -94,4 +94,8 @@ def self.shorten_provider_type(provider_type)
     @short_provider_type ||= self.class.shorten_provider_type(provider_type)
   end
+
+  def provider_type_nextcloud?
+    provider_type == ::Storages::Storage::PROVIDER_TYPE_NEXTCLOUD
+  end
```
@wielinde retaining this method in the parent `Storage` model as the use case applies in this context. E.g. in `SetAttributesService` we want to `#set_nextcloud_application_credentials_defaults` on the condition that the storage is a `provider_type_nextcloud?` ; as the attribute service is  agnostic of the different types of storages.

akabiru commented on modules/storages/app/models/storages/nextcloud_storage.rb:
```diff
@@ -27,34 +27,23 @@
+  store_attribute :provider_fields, :password, :string
+  store_attribute :provider_fields, :has_managed_project_folders, :boolean
+  store_attribute :provider_fields, :group, :string, default: PROVIDER_FIELDS_DEFAULTS[:username]
+  store_attribute :provider_fields, :group_folder, :string, default: PROVIDER_FIELDS_DEFAULTS[:username]
```
@ba1ash Updated to use stored_attributes- looks really neat! ✨

ba1ash replied:
Yep :+1:

akabiru committed:
chore[#48289] Rename stimulus controller for consistency

https://community.openproject.org/work_packages/48289

akabiru committed:
WIP: extend use of `Storages::NextcloudStorage`

akabiru committed:
fix[OP#48289]: Relax username default due to automatically managed false clearance rule

akabiru committed:
fix[OP#48289]: Infer the storage model by provider type for model specific provider fields

akabiru committed:
chore[OP#48289]: Use convenience provider_type_nextcloud?

akabiru committed:
tests[OP#48289]: Update assertion for "Password can't be blank"

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

akabiru committed:
fix[OP#48289]: Update nullable definition compliant with openapi: v3.1.0

See https://stackoverflow.com/questions/48111459/how-to-define-a-property-that-can-be-string-or-null-in-openapi-swagger

akabiru committed:
fix[OP#48289]: Update tests to use the `nextcloud_storage` factory

akabiru committed:
fix[OP#48289]: Update application password representation and scope it to nextcloud

akabiru committed:
fix[OP#48289]: Update more tests to use the `nextcloud_storage` factory

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

mereghost commented on modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb:
```diff
@@ -55,14 +55,21 @@ def new
       flash[:notice] = I18n.t(:notice_successful_create)
-      # admin_settings_storage_path is automagically created by Ruby routes.
-      redirect_to admin_settings_storage_path(@storage)
+      if OpenProject::FeatureDecisions.automatically_managed_project_folders_active?
+        if @storage.automatic_management_unspecified?
+          redirect_to new_admin_settings_storage_automatically_managed_project_folders_path(@storage)
+        else
+          redirect_to admin_settings_storage_path(@storage)
+        end
+      else
+        redirect_to admin_settings_storage_path(@storage)
+      end
```
2 of the branches end up in the same place. Do we need the nested `if`s for some reason or can we simply join then?
```suggestion
      if OpenProject::FeatureDecisions.automatically_managed_project_folders_active? && @storage.automatic_management_unspecified?
          redirect_to new_admin_settings_storage_automatically_managed_project_folders_path(@storage)
      else
        redirect_to admin_settings_storage_path(@storage)
      end
```

akabiru replied:
~Given one branch is wrapped in a feature flag, that is bound to be removed- I found it more readable in it's current form and simpler to understand- albeit minimal duplication.~

Scratch that it's actually more compact- good looking out @mereghost ! 👍🏾

mereghost commented on modules/storages/app/services/storages/storages/set_attributes_service.rb:
```diff
@@ -54,5 +76,9 @@ def derive_default_storage_name
     end
+
+    def nextcloud_storage?
+      storage.provider_type_nextcloud? && storage.is_a?(Storages::NextcloudStorage)
```
Noob question: Is there a way that the second can be true if the first isn't? 😕

akabiru replied:
Good eye, I think we can just do with the latter 👍🏾

mereghost reviewed (commented):
I don't have enough context on the feature/architecture but code wise seems good.

Just a couple of comments/questions. =)

akabiru committed:
chore[OP#48289] remove duplicate model check

Kharonus commented on frontend/src/app/spot/components/switch/switch.component.html:
```diff
@@ -5,6 +5,7 @@
   [disabled]="disabled"
   [tabindex]="tabindex"
   [ngModel]="checked"
+  [value]="checked"
```
why is that needed? `[ngModel]` is already bound to `checked`.

akabiru replied:
In this case, the component is not used in a pure angular context- the data is submitted in the context of a Rails view (HTML) form. In order to post the form data in vanilla HTML, we need the value attribute defined.

Kharonus commented on modules/storages/app/views/storages/admin/storages/automatically_managed_project_folders/edit.html.erb:
```diff
@@ -0,0 +1,102 @@
+
+    <div class="form--field-container -horizontal">
+      <label class="form--label -flex" data-storages--automatically-managed-project-folders-form-target="automaticallyManagedSwitch">
+        <%= angular_component_tag 'spot-switch',
```
We should put the stimulus target on the `spot-switch` itself, not the label wrapping it. Then we can simplify the stimulus controller.

akabiru replied:
I'm not sure the `angular_component_tag` helper supports passing data attributes atm

Kharonus commented on modules/storages/lib/api/v3/storages/storage_representer.rb:
```diff
@@ -81,6 +81,28 @@ def initialize(model, current_user:, embed_links: nil)
+             skip_parse: true,
+             skip_render: ->(represented:, **) { !represented.provider_type_nextcloud? },
+             getter: ->(represented:, **) {
+               break unless represented.provider_type_nextcloud?
```
is this break needed, if you already define `skip_render` with the same condition?

akabiru replied:
We need to guard against non-nextcloud models which do not define nextcloud provider_fields methods such as `#automatically_managed?` which is only defined in `Storages::NextcloudStorage`.

AFAICT from the unit tests `skip_render` is not exclusive of the `getter`. It seems both expressions are still evaluated; regardless of `skip_render` evaluating to `true`.

See failing tests after removing the break guard clause

```ruby
getter: ->(represented:, **) {
              #  break unless represented.provider_type_nextcloud?

               represented.automatically_managed?
             },
#
# $ docker compose exec backend-test bundle exec rspec modules/storages/spec/lib/api/v3/storages
  1) API::V3::Storages::StorageRepresenter rendering properties Automatically managed project folders with a non-Nextcloud storage does not include the property hasApplicationPassword
     Failure/Error: subject(:generated) { representer.to_json }

     NoMethodError:
       undefined method `automatically_managed?' for #<Storages::Storage id: nil, provider_type: "unknown", name: "Storage 18", host: "https://host18.example.com", creator_id: nil, created_at: nil, updated_at: nil, provider_fields: {}>
```

akabiru committed:
chore[OP#48289]: simplify conditional branching

Co-authored-by: Marcello Rocha <mereghost@users.noreply.github.com>

Kharonus commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -44,6 +44,7 @@ class Engine < ::Rails::Engine
       OpenProject::FeatureDecisions.add :storage_file_picking_select_all
       OpenProject::FeatureDecisions.add :storage_project_folders
       OpenProject::FeatureDecisions.add :managed_project_folders
+      OpenProject::FeatureDecisions.add :automatically_managed_project_folders
```
this is duplicate, isn't it? `:managed_project_folders` was already defined for the scope of anything that goes for the automatically managed folders.

akabiru replied:
IIRC we discussed and agreed to define a dedicated flag as this is in the global administration. `managed_project_folders` is for Project settings folder mode activation. Perhaps we could have named it better to be more clear. cc @wielinde

Kharonus commented on frontend/src/stimulus/controllers/dynamic/storages/automatically-managed-project-folders-form.controller.ts:
```diff
@@ -0,0 +1,88 @@
+  declare isAutomaticallyManagedValue:boolean;
+
+  spotSwitchComponent!:HTMLElement;
+  boundUpdateDisplay!:(evt:InputEvent) => (void | boolean);
```
I'd like to suggest some structural changes in the stimulus controller, that hopefully simplifies the code here:

- change signature `updateDisplay` to 

```typescript
public updateDisplay(_evt:InputEvent):void
```

- use it directly via an stimulus action on the spot switch component (if possible, needs testing)

if you want to check this together, I would gladly have a look at it. We also can do this in a follow up step.

But overall I think, this controller can be reduced by more then half.

Kharonus reviewed (changes requested):
I have some small remarks regarding the stimulus parts of the code.

ba1ash committed:
Use nextcloud_storage.automatically_managed field.

Instead of has_managed_project_folders.

ba1ash committed:
Fix group_folder_properties_sync_service spec.

Remove non existent has_managed_project_folders field.

akabiru committed:
Merge branch 'dev' into implementation/48289-add-automatic-project-folder-management-settings-for-storages

akabiru committed:
chore[OP#48289] user trait for automatic management factory

Kharonus committed:
[#48289] improved controller for app password form (#13070)

ba1ash commented on modules/storages/app/models/storages/nextcloud_storage.rb:
```diff
@@ -27,22 +27,24 @@
+  store_attribute :provider_fields, :automatically_managed, :boolean
+  store_attribute :provider_fields, :username, :string
+  store_attribute :provider_fields, :password, :string
+  store_attribute :provider_fields, :group, :string, default: PROVIDER_FIELDS_DEFAULTS[:username]
+  store_attribute :provider_fields, :group_folder, :string, default: PROVIDER_FIELDS_DEFAULTS[:username]
```
Newly generated accessors behaves differently. A specified default value is returned only when the record is new. So, we have to decide how to handle the case, for instance, with existing records where `group` and `group_folder` have not been saved to the database. And we should make sure that when we create new records using the form `group` and `group_folder` are saved to the database or use something like previous implementation with the following logic: `provider_fields['group'] || ''OpenProject'`

akabiru replied:
Good catch, reinstating a runtime default as you proposed sounds safer. It seems store_attribute is mainly useful for typecasting.

akabiru replied:
Added https://github.com/opf/openproject/pull/12721/commits/a78ca272136d2a6cadadc02a3de6ed47d491f1dd

akabiru committed:
fix[OP#48289]: Redefine group_{folder} runtime default values to account for existing records

akabiru committed:
fix[OP#48289]: Do not persist system provider fields defaults

ba1ash commented on modules/storages/app/services/storages/storages/set_attributes_service.rb:
```diff
@@ -36,12 +36,31 @@ def set_default_attributes(_params)
       storage.name ||= derive_default_storage_name
     end
 
+    def set_default_provider_fields(_params)
+      set_nextcloud_application_credentials_defaults if nextcloud_storage?
+    end
+
```
```suggestion
```

ba1ash commented on modules/storages/app/services/storages/storages/set_attributes_service.rb:
```diff
@@ -36,12 +36,31 @@ def set_default_attributes(_params)
 
+    def set_attributes(params)
+      super(params)
+      set_default_provider_fields(params)
```
```suggestion
      remove_nextcloud_credentials if nextcloud_storage?
```

ba1ash commented on modules/storages/app/services/storages/storages/set_attributes_service.rb:
```diff
@@ -36,12 +36,31 @@ def set_default_attributes(_params)
       storage.host = storage.host&.gsub(/\/+$/, '')
     end
 
+    def set_nextcloud_application_credentials_defaults
+      # Do not overwrite if has never been set.
+      # E.g. when setting up a new storage for the first time, passthrough, credentials are set in a later stage.
+      return if storage.automatic_management_unspecified?
+
+      unless storage.automatically_managed?
+        %w[username password].each { |field| storage.provider_fields.delete(field) }
+      end
+    end
+
```
```suggestion
    def remove_nextcloud_credentials
      # Do not overwrite if has never been set.
      # E.g. when setting up a new storage for the first time, passthrough, credentials are set in a later stage.
      return if storage.automatic_management_unspecified?

      unless storage.automatically_managed?
        %w[username password].each { |field| storage.provider_fields.delete(field) }
      end
    end

```
